control-plane-operator/controllers/azureprivateendpoint/azureprivateendpoint_controller.go

```diff
@@ -0,0 +1,327 @@
+			controllerutil.RemoveFinalizer(azurePrivateEndpoint, finalizer)
+			if err := r.Patch(ctx, azurePrivateEndpoint, client.MergeFrom(originalObj)); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to remove finalizer: %w", err)
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+
+	if len(hcpList.Items) > 1 {
+		return ctrl.Result{}, fmt.Errorf("unexpected number of HostedControlPlanes in namespace, expected: 1, actual: %d", len(hcpList.Items))
+	}
+	hcp := &hcpList.Items[0]
+
+	if isPaused, duration := util.IsReconciliationPaused(log, hcp.Spec.PausedUntil); isPaused {
+		log.Info("Reconciliation paused", "pausedUntil", *hcp.Spec.PausedUntil)
+		return ctrl.Result{RequeueAfter: duration}, nil
+	}
+
+	// Return early if deleted
+	if !azurePrivateEndpoint.DeletionTimestamp.IsZero() {
+		if !controllerutil.ContainsFinalizer(azurePrivateEndpoint, finalizer) {
+			// If we previously removed our finalizer, don't delete again and return early
+			return ctrl.Result{}, nil
+		}
+		if err := r.delete(ctx, azurePrivateEndpoint, hcp); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to delete resource: %w", err)
+		}
+
+		if controllerutil.ContainsFinalizer(azurePrivateEndpoint, finalizer) {
+			controllerutil.RemoveFinalizer(azurePrivateEndpoint, finalizer)
+			if err := r.Patch(ctx, azurePrivateEndpoint, client.MergeFrom(originalObj)); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to remove finalizer: %w", err)
+			}
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// Ensure the azurePrivateEndpoint has a finalizer for cleanup
+	if !controllerutil.ContainsFinalizer(azurePrivateEndpoint, finalizer) {
+		controllerutil.AddFinalizer(azurePrivateEndpoint, finalizer)
+		if err := r.Patch(ctx, azurePrivateEndpoint, client.MergeFrom(originalObj)); err != nil {
+			if apierrors.IsConflict(err) {
+				return ctrl.Result{Requeue: true}, nil
+			}
+			return ctrl.Result{}, fmt.Errorf("failed to add finalizer: %w", err)
+		}
+	}
+
+	// Reconcile the AzurePrivateEndpoint
+	originalObj = azurePrivateEndpoint.DeepCopy()
+	if err := r.reconcileAzurePrivateEndpoint(ctx, azurePrivateEndpoint, hcp); err != nil {
+		meta.SetStatusCondition(&azurePrivateEndpoint.Status.Conditions, metav1.Condition{
+			Type:    string(hyperv1.AzurePrivateEndpointAvailable),
+			Status:  metav1.ConditionFalse,
+			Reason:  hyperv1.AzureErrorReason,
+			Message: err.Error(),
+		})
+		if err := r.Status().Patch(ctx, azurePrivateEndpoint, client.MergeFrom(originalObj)); err != nil {
+			return ctrl.Result{}, err
+		}
+
+		return ctrl.Result{}, err
+	}
+
+	meta.SetStatusCondition(&azurePrivateEndpoint.Status.Conditions, metav1.Condition{
+		Type:    string(hyperv1.AzurePrivateEndpointAvailable),
+		Status:  metav1.ConditionTrue,
+		Reason:  hyperv1.AsExpectedReason,
+		Message: "",
+	})
+	if err := r.Status().Patch(ctx, azurePrivateEndpoint, client.MergeFrom(originalObj)); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	log.Info("reconcilation complete")
```

Comment:
Should this be more specific like `reconciliation complete - azure private endpoint`